### PR TITLE
Form Cleanup (Volume 2)

### PR DESCRIPTION
**Pull request by bastianallgeier**
### Summary of changes

- `Form::$fields` is no longer optional. 
- Replace `empty()` calls
- Field Siblings refactoring
  - Rename `Field::$formFields` to `Field::$siblings` for more consistency 
  - Added a new `Field::siblings()` method and deprecated `Field::formFields()`
  - Added the `HasSiblings` trait to align `Field` with `FieldClass` This also needed a new protected `Field::siblingsCollection` method
- Sorted all methods in FieldTest alphabetically and added missing test methods
- Renamed `Field::save()` to `Field::isSaveable()` to align it with the `FieldClass` 
- New `Field::isEmptyValue()` method to align it with `FieldClass` Passing an argument to `Field::isEmpty()` is now deprecated. 
- New Trait classes to remove redundancies between `Field` and `FieldClass`
  - `Kirby\Form\Mixin\Api`
  - `Kirby\Form\Mixin\Model`
  - `Kirby\Form\Mixin\Validation`
  - `Kirby\Form\Mixin\When`
- New `$model` argument for the `Kirby\Form\Fields` constructor. This will automatically inject the model into all fields.

### Reasoning

The final goal of this cleanup is to get to a point where we can construct a form and fill it with values afterwards: 

```php
$form = new Form([
  'fields' => []
]);

$form->fill([
  'a' => 'A',
  'b' => 'B',
]);
```

Why is this needed? For our models, we want to be able to set the original values for the form fields and then overwrite changed fields in a second fill call. 

```php
$form = new Form([
  'fields' => [], 
  'values' => [
    // original values
  ]
]);

$form->fill([
  // changed values
]);
```

This should not cause the entire form and every field to be reconstructed and every property to be reevaluated. Otherwise we will create massive performance issues. 

But our Form class is full of spaghetti code. We also have a lot of inconsistencies between `Field` and `FieldClass`. The combination of those "dirty" parts lead to a situation where it is basically impossible to create such a `Form::fill` method. 

### Next steps

This cleanup is not done yet. In a next step, the Form class needs to be refactored more to get rid of a very big foreach loop in the constructor, which handles all kind of weird edge cases for fields and values. We also need to build the final ::fill method and make sure that all our tests still pass and we don't introduce any form regressions. 

### Deprecated

- `Field::save()`: use `Field::isSaveable()` instead
- `Field::formFields()`: use `Field::siblings()` instead
- Passing a value to `Field::isEmpty($value)` is deprecated. Use `Field::isEmptyValue($value)` instead to check if a value qualifies as empty for this field type. `Field::isEmpty()` should be reserved to check if the current Field value is empty. 

### Breaking changes

- `Kirby\Form\Field::$formFields` has been renamed to `Kirby\Form\Field::$siblings`
- The `$formField` argument in `Kirby\Form\Field::factory()` has been renamed to `$siblings`
- `Kirby\Form\Form::fields()` does no longer return null but an empty Fields collection if there are no fields. 
- `Kirby\Form\Field::validate()` returns an array of errors instead of void. The validate method is protected and this should not really qualify as a public breaking change, but it's still worth noting it here. 

## Ready?

- [x] In-code documentation (wherever needed)
- [x] Unit tests for fixed bug/feature
- [x] Tests and CI checks all pass

### For review team

- [ ] Add changes & docs to release notes draft in Notion

### CODE REVIEW

**Comment by bastianallgeier** [comment]
This is a big one. @getkirby/backend let me know if this is still ok to review, if you need anything else from me in the description or if I should break it down into smaller steps. I tried to do that with the commits and keep each one mostly focused on one particular task. I hope it helps.

File: src/Form/Field.php [hunk 1]
```diff
@@ -50,7 +52,7 @@ class Field extends Component
 	public function __construct(
 		string $type,
 		array $attrs = [],
-		Fields|null $formFields = null
+		Fields|null $siblings = null
```
Comment (distantnative) [inline src/Form/Field.php:55]: `Todo:` We need to list the argument name change as breaking change.

File: src/Form/Field.php [hunk 2]
```diff
@@ -290,16 +274,16 @@ public function drawers(): array
 	public static function factory(
 		string $type,
 		array $attrs = [],
-		Fields|null $formFields = null
+		Fields|null $siblings = null
```
Comment (distantnative) [inline src/Form/Field.php:277]: `Todo:` List as breaking change.

File: src/Form/Field.php [hunk 3]
```diff
@@ -323,23 +307,11 @@ public function fill(mixed $value): static
 	}
 
 	/**
-	 * Parent collection with all fields of the current form
+	 * @deprecated Use `::siblings() instead
```
Comment (distantnative) [inline src/Form/Field.php]:
```suggestion
	 * @deprecated  5.0.0 Use `::siblings() instead
```

File: src/Form/Field.php [hunk 4]
```diff
@@ -352,6 +324,7 @@ public function isDisabled(): bool
 
 	/**
 	 * Checks if the field is empty
+	 * @deprecated Passing arguments is deprecated. Use `::isEmptyValue()` instead to check for
```
Comment (distantnative) [inline src/Form/Field.php]:
```suggestion
	 * @deprecated 5.0.0 Passing arguments is deprecated. Use `::isEmptyValue()` instead.
```

File: src/Form/Field.php [hunk 5]
```diff
@@ -360,6 +333,14 @@ public function isEmpty(mixed ...$args): bool
 			default => $args[0]
 		};
 
+		return $this->isEmptyValue($value);
+	}
+
+	/**
+	 * Checks if the given value is considered empty
+	 */
```
Comment (distantnative) [inline src/Form/Field.php:343]:
```suggestion
   * @since 5.0.0
	 */
```

File: src/Form/Field.php [hunk 6]
```diff
@@ -392,77 +365,51 @@ public function isRequired(): bool
...
+			return $this->options['api']->call($this);
+		}
+
+		return [];
 	}
 
 	/**
-	 * Returns the parent model
+	 * Checks if the field is saveable
+	 * @deprecated Use `::isSaveable()` instead
```
Comment (distantnative) [inline src/Form/Field.php]:
```suggestion
	 * @deprecated 5.0.0 Use `::isSaveable()` instead
```

File: src/Form/Mixin/When.php [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+			$this->isSaveable() === false ||
+			$this->isRequired() === false ||
+			$this->isEmpty() === false ||
+			$this->isActive() === false
+		) {
+			return false;
+		}
+
+		return true;
+	}
```
Comment (distantnative) [inline src/Form/Mixin/When.php:63]: Really unsure if this shouldn't be in the validation mixin. It's somewhat part of both.

Comment (bastianallgeier) [reply src/Form/Mixin/When.php:63]: I'm working on additional steps with a new Value mixin. I think it should actually go in there later. Should I move it back to the classes for this PR or do you think it's fine to move it to a different Mixin later?

Comment (distantnative) [reply src/Form/Mixin/When.php:63]: It's fine to have it in the mixing and move it later.

**Comment by distantnative** [comment]
@bastianallgeier something broke but it seems like already on `v5/changes/develop` - I still think we need to fix it here/first as it makes testing this PR hard:

1. Check out lab
2. Edit `site/blueprints/pages/text-field.yml`
3. Add when conditions to the two validation fields:

```yml
     required:
        type: text
        required: true
        when:
          text: "foo"
      minlength:
        type: text
        minlength: 10
        when:
          text: "foo"
```

4. Go the page in channel. Fields will be hidden as long as the `text` field value isn't `foo` ☑️
5. Then you can also type in anything but `foo` and save, it gets saved as the validations of the hidden fields are rightfully ignored ☑️
6. Type in `foo` to the text field
7. The fields in the validations tab show up  ☑️
8. Try to save
9. No error message is shown instead of an error message that the `required` field has no value but is required ❌

**Comment by bastianallgeier** [comment]
@distantnative I started by adding a new when query test setup to the sandbox. https://github.com/getkirby/sandbox/commit/76bd69b848b9a63473d39995967714c4a1ed7f7c

**Comment by bastianallgeier** [comment]
Turns out: we don't run any form of validation on v5/changes/develop so far 🙈

It's really not related to this PR though.

**Comment by distantnative** [comment]
@bastianallgeier 🙈

But I think we need to fix it first or else we won't know if this PR leaves everything else functioning. I know you don't want this to become stuck but it's hard to test without.

**Comment by bastianallgeier** [comment]
I agree. I will create a new PR to fix this.

**Review by distantnative — APPROVED** [review]
